pkg/cache/cache.go

```diff
@@ -0,0 +1,97 @@
+package cache
+
+import (
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"go.uber.org/zap"
+)
+
+type Cache struct {
+	cacheDir string
+}
+
+func New(rootDir string) (*Cache, error) {
+	cacheDir := filepath.Join(rootDir, "cache")
+	if err := os.MkdirAll(cacheDir, os.ModePerm); err != nil {
+		return nil, fmt.Errorf("creating a cache directory: %w", err)
+	}
+
+	return &Cache{cacheDir: cacheDir}, nil
+}
+
+func (cache *Cache) Get(fileIdentifier string) (path string, err error) {
```

Comment:
The argument structure is intentionally left out to the caller. It may be a filename, a URL or a custom string (e.g. `v1.28.4+rke2r1/rke2-images-core.linux-amd64.tar.zst` in the case of Kubernetes artefact downloader). This provides the most flexibility for now and we can always revisit it later if a more structured approach is required.

The same thing applies for cache wrappers i.e. if we want to reuse the caching elsewhere and the `copyFromCache` -> `download` flow is repeated we can always abstract it out.